### PR TITLE
chore: created and tested DelegatedAccess entity

### LINKED ISSUE #36: Create Entity PoAAccess
#### Overview:
`PoAAccess` (Power of Attorney Access) is an entity designed to represent access given by an account holder or manager to another person or entity through a Power of Attorney (PoA). This allows the designated person (the attorney-in-fact or agent) to act on behalf of the account holder in a wide range of financial operations. PoAAccess is highly flexible and can be used for anything from day-to-day transactions to managing large financial activities. It can also be used to implement open banking access, as it grants the necessary permissions for third-party interactions while maintaining the account holder’s control.

A PoAAccess role is defined by the scope of the powers granted, meaning it can be narrowly focused or very broad, depending on the needs of the account holder. This makes PoAAccess a versatile access entity in banking systems, especially for delegation of tasks that require the management of finances, asset transfers, and decision-making.

#### Key Characteristics:

1. **Delegated Access Through Power of Attorney (PoA):**
   - `PoAAccess` allows the account holder or a manager to grant another individual or entity access to the account under a Power of Attorney (PoA). This access can include various permissions, such as:
     - Managing account balances and transactions.
     - Approving payments or transferring funds.
     - Opening new accounts or closing existing ones.
     - Handling investment-related tasks.
     - Accessing financial information for the account.
   - The exact scope of this access is determined by the account holder or manager, making it a highly customizable form of access.

2. **Granularity of Access:**
   - `PoAAccess` is highly granular, allowing the account holder or manager to define specific tasks or operations that the attorney-in-fact can perform. For example:
     - An account holder may grant PoAAccess to an accountant to manage tax-related transactions but restrict the ability to make transfers or withdrawals.
     - Alternatively, the account holder can give broader powers to a family member or business partner, allowing them to fully manage the account during a period of incapacity or absence.
   - This flexibility makes PoAAccess an essential tool for delegating complex or sensitive financial responsibilities.

3. **Alignment with Open Banking Access:**
   - `PoAAccess` can be leveraged to implement open banking access, where an account holder delegates part of their authority to a third-party provider (TPP). This is similar to `ThirdPartyAccess` but is structured under legal PoA rules, allowing for broader control. For example:
     - **Account Information Access:** An attorney-in-fact can retrieve account details or transaction histories from the bank.
     - **Payment Initiation Access:** The attorney-in-fact may initiate payments or transfers, including recurring payments, on behalf of the account holder.
     - **Confirmation of Funds:** Similar to open banking’s "confirmation of funds" consent, PoAAccess can allow the attorney-in-fact to verify available funds for specific purposes (e.g., large purchases or loan guarantees).
   - Using PoAAccess for open banking provides legal grounding and transparency, ensuring that the third party acts within the boundaries set by the account holder.

4. **Types of Power of Attorney:**
   - `PoAAccess` can represent various forms of Power of Attorney, depending on the account holder’s needs:
     - **General PoA:** Grants broad access to manage the account and make financial decisions.
     - **Limited PoA:** Restricts access to specific tasks or transactions, such as making a single payment or handling a specific set of financial operations.
     - **Durable PoA:** Remains in effect even if the account holder becomes incapacitated, giving the attorney-in-fact ongoing control over the account.
     - **Springing PoA:** Only becomes effective under certain conditions, such as if the account holder becomes incapacitated or is otherwise unable to manage their own affairs.
   - Each type of PoA offers different levels of control, depending on the account holder’s preferences and circumstances.

5. **No Access Management:**
   - Similar to other access types like `AgentAccess`, `PoAAccess` does not grant the power to manage access for other users. The attorney-in-fact cannot add, modify, or revoke access for other entities associated with the account.
   - They can only perform the specific financial tasks or operations defined in the Power of Attorney, ensuring that the overall control remains with the account holder or account manager.

6. **Weight:**
   - The `weight` of `PoAAccess` can vary depending on the breadth of the authority granted under the Power of Attorney:
     - **0.5 - 1.0:** For broad or general powers where the attorney-in-fact has substantial control over the account’s operations.
     - **0.1 - 0.4:** For limited PoA, where the attorney-in-fact has restricted or specific access (e.g., only for tax filings or certain types of transactions).
   - The weight reflects the level of influence the attorney-in-fact has over the account and ensures that access is proportional to the delegated responsibilities.

7. **Status:**
   - The status of `PoAAccess` can change based on the conditions of the Power of Attorney:
     - **Active:** The PoA is currently in effect, and the attorney-in-fact can perform the tasks defined in the Power of Attorney.
     - **Suspended:** Access is temporarily suspended, often due to legal reasons or by decision of the account holder or manager.
     - **Revoked:** The account holder or manager has withdrawn the Power of Attorney, cutting off the attorney-in-fact’s access to the account.
     - **Expired:** The PoA was only valid for a limited time or a specific transaction, and the access has naturally expired after fulfilling its purpose.
   - These status changes ensure that `PoAAccess` remains tightly controlled and can be dynamically adjusted based on the account holder’s needs.

8. **Conditions and Restrictions:**
   - The account holder or manager can impose specific conditions on `PoAAccess` to further restrict the scope of the attorney-in-fact’s authority:
     - **Monetary Limits:** The attorney-in-fact may only be able to initiate transactions below a certain amount or within specific financial thresholds.
     - **Temporal Limits:** The PoA may only be valid for a certain period or for specific circumstances, such as during the account holder’s absence.
     - **Operational Limits:** The attorney-in-fact may only perform specific operations (e.g., tax payments, managing a particular business expense) and be restricted from others (e.g., making personal withdrawals).

#### Example Workflow for PoAAccess:

1. **Handling Business Finances with General PoA:**
   - A business owner grants a trusted partner `PoAAccess` with general powers to manage the company’s finances while they are abroad for several months.
   - The partner can initiate payments to suppliers, manage payroll, and transfer funds between accounts, but cannot change access permissions for other roles.
   - This broad PoA ensures that the company’s finances remain operational during the owner’s absence.

2. **Limited PoA for Tax Filings:**
   - An account holder grants their accountant `PoAAccess` with limited powers to access transaction histories and initiate payments related to tax filings.
   - The accountant can review the account’s financial data for tax purposes and pay outstanding tax liabilities but cannot make unrelated transactions or alter account settings.
   - This limited PoA ensures that the accountant only has access to the information and operations necessary for fulfilling their specific task.

3. **Revoking PoA After a Transaction:**
   - An elderly account holder grants their adult child `PoAAccess` to manage the sale of a property and handle the associated financial transactions.
   - Once the sale is complete, the account holder revokes the PoA to prevent further access by the child, ensuring that the control returns entirely to the account holder.

#### Key Considerations for PoAAccess:

- **Legal Validity and Scope:** The scope of a Power of Attorney is legally defined, and the permissions granted through `PoAAccess` must comply with legal and regulatory frameworks. This ensures that the attorney-in-fact operates within clearly defined boundaries.
- **Revocability and Expiration:** The account holder or manager must be able to revoke or modify the PoA at any time, ensuring that access can be quickly withdrawn if circumstances change or if misuse is suspected.
- **Accountability:** All actions taken under `PoAAccess` should be logged, ensuring transparency and providing a clear audit trail of the attorney-in-fact’s activities.

#### Scenarios for Revocation or Expiry:

- **Completion of Task:** Once the specific task for which the PoA was granted has been completed, the access can be revoked or allowed to expire naturally.
- **Change in Account Holder’s Circumstances:** If the account holder regains their capacity to manage the account (e.g., after recovering from illness), they may choose to revoke the PoA.
- **Misuse of Access:** If the attorney-in-fact is found to have misused the granted permissions, the account holder or manager can immediately revoke `PoAAccess` to prevent further unauthorized actions.

#### Example Use Cases:

- **Managing Elderly or Incapacitated Individuals’ Finances:** A family member may be granted PoA to manage an elderly relative’s account, ensuring their bills are paid, and their financial matters are handled properly without full account ownership being transferred.
- **Business Partnerships:** In a business, one partner might grant PoA to another for managing financial operations while traveling or unavailable, ensuring that the business can continue operating smoothly.
- **Complex Financial Transactions:** For large transactions, such as real estate purchases or legal settlements, a legal representative might be granted PoA to execute the deal on behalf of the account holder, ensuring that everything is handled efficiently.

### PR DESCRIPTION
**Pull request by Koufan-De-King**
Created

- PointOfAttorney access class,
- a repository for it's eventual connection to the database
and wrote a basic test using JUnit.

### CODE REVIEW

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/PoAAccess.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    }
+
+    public boolean isActive() {
+        LocalDateTime now = LocalDateTime.now();
+        return now.isAfter(validFrom) && now.isBefore(validUntil);
+    }
+}
+
+
```
Comment (NkwaTambe) [inline ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/domain/PoAAccess.java]: Please use the `Getters and Setter annotation properly`

you have `lombok` already, make good use of it

you are doing a great job but your code can be done in a better way.

you can refer to these : https://medium.com/@wandercosta/lombok-getters-and-setters-10ce392188f6

**Review by NkwaTambe — CHANGES_REQUESTED** [review]
Good job @Koufan-De-King 

please look through  your code again and ticket description..

you can reconsider refactoring your code if you want

**Comment by NkwaTambe** [comment]
Please add ticket description

**Comment by Koufan-De-King** [comment]
> Please add ticket description

Done

File: ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/repository/DelegatedAccessRepository.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface DelegatedAccessRepository extends JpaRepository<DelegatedAccess, String> {
+
+    List<DelegatedAccess> findByAccountId(String accountId123);
+}
```
Comment (NkwaTambe) [inline ledgers-bank-account-access-management/ledgers-bank-account-access-repository/src/main/java/de/adorsys/ledgers/baam/db/repository/DelegatedAccessRepository.java]: `accountId123`????


were is these from  ?

**Review by NkwaTambe — APPROVED** [review]

**Review by Arielpetit — APPROVED** [review]
Looks good

**Pull request merged**